e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,14 +223,29 @@ func (pf *PortForwarder) start() error {
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
 	go func() {
 		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
+			// create an inner context for the watch so we can cancel threads between iterations
+			// the subprocess will be killed inside portForwardToDeployment
+			watchCtx, watchCancel := context.WithCancel(pf.ctx)
+			// restart the port-forward if:
+			// - the subprocess is killed
+			// - the pod is restarted
+			restartChan := pf.signalOnFailure(
+				func() error {
+					return pf.cmd.Wait()
+				},
+				func() error {
+					return pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+						testwatcher.WatchTimeout(0), testwatcher.WatchContext(watchCtx))
+				},
+			)
+			select {
+			case <-pf.ctx.Done():
```

Comment:
why not just put this blocking on ctx.Done() as one of the parallel functions?

Reply:
We would need another signal to tell it to exit instead of restart (e.g. `doneChan`). At that point we're creating another signal which is pretty much just a proxy for `ctx.Done`, so I feel that may just be making things more complicated